File: python/test/mod_test_normalization_pipeline.py
"""
Test suite for tox_normalization_pipeline (Python)
Mirrors R/Fortran tests: basic, edge case, pipeline vs manual
"""
import numpy as np
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from tensoromics_functions import tox_normalization_pipeline, tox_normalize_by_std_dev, tox_quantile_normalization, tox_calculate_tissue_averages, tox_log2_transformation
from test_helpers import run_all_tests


def test_basic():
    n_genes, n_tissues, n_grps = 10, 6, 2
    input_matrix = np.empty((n_tissues, n_genes), dtype=np.float64, order="F")
    for i in range(n_genes):
        for j in range(n_tissues):
            input_matrix[j, i] = (i + 1) * 2.0 + (j + 1) * 0.5
    group_s = np.array([1,4], dtype=np.int32)
    group_c = np.array([3,3], dtype=np.int32)
    result = tox_normalization_pipeline(input_matrix, group_c, span=0.75, degree=2, use_quantile=1)
    assert np.all(~np.isnan(result)), "NaN in output"
    assert np.all(result >= 0), "Negative values in output"
    assert result.shape == (n_grps, n_genes), f"Output shape mismatch: {result.shape}"


def test_edge_case():
    n_genes, n_tissues, n_grps = 10, 6, 2
    input_matrix = np.zeros((n_genes, n_tissues), dtype=np.float64)
    group_s = np.array([1,4], dtype=np.int32)
    group_c = np.array([3,3], dtype=np.int32)
    try:
        tox_normalization_pipeline(input_matrix, group_c, span=0.75, degree=2, use_quantile=1)
        assert False, "Expected error for zero-variance input"
    except Exception as err:
        pass


def test_pipeline_vs_manual():
    """Compare normalization pipeline with manual stepwise operations using LOESS"""
    # print("")
    # print("=== Test: test_pipeline_vs_manual ===")
    n_genes, n_tissues, n_grps = 10, 6, 2
    col_names = [
        "muscle_dietM_1", "muscle_dietM_2", "muscle_dietM_3",
        "muscle_dietP_1", "muscle_dietP_2", "muscle_dietP_3",
    ]
    input_matrix = np.empty((n_tissues, n_genes), dtype=np.float64, order="F")
    for i in range(n_genes):
        for j in range(n_tissues):
            input_matrix[j, i] = (i + 1) * 2.0 + (j + 1) * 0.5
    group_s = np.array([1,4], dtype=np.int32)
    group_c = np.array([3,3], dtype=np.int32)
    span = 0.75
    degree = 2

    # Test WITH quantile normalization
    result_pipeline_quantile = tox_normalization_pipeline(
        input_matrix.copy(), group_c, span=span, degree=degree, use_quantile=1
    )

    # Manual with quantile: std_dev → quantile → avg → log2
    temp1 = tox_normalize_by_std_dev(input_matrix.copy(), span=span, degree=degree)
    temp2 = tox_quantile_normalization(temp1)
    temp3 = tox_calculate_tissue_averages(temp2, group_c)
    result_manual_quantile = tox_log2_transformation(temp3)

    # Test WITHOUT quantile normalization
    result_pipeline_no_quantile = tox_normalization_pipeline(
        input_matrix.copy(), group_c, span=span, degree=degree, use_quantile=0
    )

    # Manual without quantile: std_dev → avg → log2
    temp1_no_q = tox_normalize_by_std_dev(input_matrix.copy(), span=span, degree=degree)
    temp2_no_q = tox_calculate_tissue_averages(temp1_no_q, group_c)
    result_manual_no_quantile = tox_log2_transformation(temp2_no_q)

    # Verify shapes
    assert result_pipeline_quantile.shape == (n_grps, n_genes), 1
    assert result_pipeline_no_quantile.shape == (n_grps, n_genes), 2
    assert result_manual_quantile.shape == (n_grps, n_genes), 3
    assert result_manual_no_quantile.shape == (n_grps, n_genes), 4

    # Pipeline should match manual operations
    assert np.allclose(result_pipeline_quantile, result_manual_quantile, rtol=1e-10), \
        "Pipeline with quantile should match manual stepwise operations"
    assert np.allclose(result_pipeline_no_quantile, result_manual_no_quantile, rtol=1e-10), \
        "Pipeline without quantile should match manual stepwise operations"

    # No NaN values
    assert np.all(~np.isnan(result_pipeline_quantile)), "NaN in pipeline output (quantile)"
    assert np.all(~np.isnan(result_pipeline_no_quantile)), "NaN in pipeline output (no quantile)"

    # Results should differ when quantile option changes
    assert np.any(np.abs(result_pipeline_quantile - result_pipeline_no_quantile) > 1e-12), \
        "Quantile and non-quantile results should differ"


if __name__ == '__main__':
    run_all_tests(globals().values())

File: r/test/mod_test_normalization_pipeline.R
# Comprehensive R test suite for normalization_pipeline (mirrors Fortran unit tests)
source("r/load_tensor_omics.R")
source("r/test_helpers.R")

test_basic <- function() {
  n_genes <- 10; n_tissues <- 6; n_grps <- 2
  input_matrix <- matrix(0, ncol=n_genes, nrow=n_tissues)
  for (i in 1:n_genes) {
    for (j in 1:n_tissues) {
      input_matrix[j, i] <- i + j * 0.25
    }
  }
  rownames(input_matrix) <- c(
    "muscle_dietM_1", "muscle_dietM_2", "muscle_dietM_3",
    "muscle_dietP_1", "muscle_dietP_2", "muscle_dietP_3"
  )
  colnames(input_matrix) <- paste0("gene", 1:n_genes)
  group_s <- c(1L,4L); group_c <- c(3L,3L)
  result <- normalization_pipeline(input_matrix, group_c, span = 0.75, degree = 2, use_quantile = TRUE)
  assert_true(all(!is.na(result)))
  assert_true(all(result >= 0))
  assert_true(dim(result)[2] == n_genes && dim(result)[1] == n_grps)
}

test_edge_case <- function() {
  n_genes <- 10; n_tissues <- 6; n_grps <- 2
  input_matrix <- matrix(0, ncol=n_genes, nrow=n_tissues)
  rownames(input_matrix) <- c(
    "muscle_dietM_1", "muscle_dietM_2", "muscle_dietM_3",
    "muscle_dietP_1", "muscle_dietP_2", "muscle_dietP_3"
  )
  colnames(input_matrix) <- paste0("gene", 1:n_genes)
  group_s <- c(1L,4L); group_c <- c(3L,3L)

  err <- tryCatch({
    normalization_pipeline(input_matrix, group_c, span = 0.75, degree = 2, use_quantile = TRUE)
    NULL
  }, error = function(e) e)

  assert_true(!is.null(err))
}

test_pipeline_vs_manual <- function() {
  n_genes <- 10; n_tissues <- 6; n_grps <- 2
  input_matrix <- matrix(0, ncol=n_genes, nrow=n_tissues)
  for (i in 1:n_genes) {
    for (j in 1:n_tissues) {
      input_matrix[j, i] <- i * 2 + j * 0.5
    }
  }
  rownames(input_matrix) <- c(
    "muscle_dietM_1", "muscle_dietM_2", "muscle_dietM_3",
    "muscle_dietP_1", "muscle_dietP_2", "muscle_dietP_3"
  )
  colnames(input_matrix) <- paste0("gene", 1:n_genes)
  group_s <- c(1L,4L); group_c <- c(3L,3L)

  buf_stddev <- normalize_by_std_dev(input_matrix)
  # quantile_normalization returns the rank means alongside the matrix
  buf_quant <- quantile_normalization(buf_stddev)$normalized_expr
  dimnames(buf_quant) <- dimnames(input_matrix)
  buf_avg <- calc_tiss_avg(group_c, buf_quant)
  manual_out <- log2_transformation(as.matrix(buf_avg))

  dimnames(buf_stddev) <- dimnames(input_matrix)
  buf_avg_no_quant <- calc_tiss_avg(group_c, buf_stddev)
  manual_out_no_quant <- log2_transformation(as.matrix(buf_avg_no_quant))

  result <- normalization_pipeline(input_matrix, group_c, span = 0.75, degree = 2, use_quantile = TRUE)
  assert_equal_numeric(as.vector(result), as.vector(manual_out), msg="test_pipeline_vs_manual")

  result_no_quant <- normalization_pipeline(input_matrix, group_c, span = 0.75, degree = 2, use_quantile = FALSE)
  assert_equal_numeric(as.vector(result_no_quant), as.vector(manual_out_no_quant), msg="test_pipeline_vs_manual_no_quant")
}

run_all_tests()